File: haskell/2023maj/rgb.hs
import System.IO
import Data.List (findIndex)
import Data.Maybe (fromJust)

type Pixel = (Int, Int, Int)

toPixels:: [Int] -> [Pixel]
toPixels [] = []
toPixels (x:y:z:xs) = (x, y, z): toPixels xs

toInt :: [String] -> [Int]
toInt = map read

-- 5. Feladat
hatar :: [Pixel] -> Int -> Bool
hatar [] _ = False
hatar [_] _ = False
hatar (x:y:xs) elteres
    | abs (blue x - blue y) > elteres = True
    | otherwise = hatar (y:xs) elteres
        where
            blue (_, _, z) = z


main :: IO()
main = do
    -- 1. Feladat: Beolvasás
    handle <- openFile "../../input/2023maj/kep.txt" ReadMode
    contents <- hGetContents handle
    let pixels = map (toPixels . toInt . words) $ lines contents

    -- 2. Feladat
    putStrLn "2. feladat:\nKérem egy képpont adatait!"
    putStrLn "Sor: "
    sor <- readLn :: IO Int
    putStrLn "Oszlop: "
    oszlop <- readLn :: IO Int
    putStrLn ("A képpont színe RGB" ++ show (pixels !! (sor - 1) !! (oszlop - 1)))

    -- 3. Feladat
    putStr "3. feladat:\nA világos képpontok száma: "
    print (length $ filter (\(x,y,z) -> x + y + z > 600) $ concat pixels)

    -- 4. Feladat
    putStr "4. feladat:\nA legsötétebb pont RGB összege: "
    let legkisebb = minimum $ map (\(x,y,z) -> x + y + z) $ concat pixels
    print legkisebb

    putStrLn "A legsötétebb pixelek színe:"
    mapM_ print $ filter (\(x,y,z) -> x+y+z == legkisebb) $ concat pixels

    -- 5. Feladat a 'main' fölött

    -- 6. Feladat
    putStrLn "6. feladat:\nA felhő legfelső sora:"
    print (1 + fromJust (findIndex (`hatar` 10) pixels))

    putStrLn "A felhő legalsó sora:"
    print (length pixels - fromJust (findIndex (`hatar` 10) $ reverse pixels))


    hClose handle

File: rust/2023maj/rgb.rs
use std::{fs, env, io};

#[derive(Debug)]
struct RGB (i32, i32, i32);

fn main() {
    // 1. Feladat: beolvasás
    let input_file = env::current_dir().unwrap().join("../../input/2023maj/kep.txt");

    let pixels: Vec<Vec<RGB>> = fs::read_to_string(input_file).unwrap()
        .trim()
        .split('\n')
        .map(|line| {
            let l: Vec<&str> = line.trim().split(' ').collect();
            l.chunks(3)
                .map(|chunk| {
                    RGB(
                        chunk[0].parse().unwrap(),
                        chunk[1].parse().unwrap(),
                        chunk[2].parse().unwrap()
                    )
                }).collect()
        }).collect();

    // 2. Feladat
    println!("2. feladat:\nKérem egy képpont adatait!");

    println!("Sor: ");
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let sor: usize = input.clone().trim().parse().unwrap();

    println!("Oszlop: ");
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let oszlop: usize = input.clone().trim().parse().unwrap();

    println!("A képpont színe {:?}", pixels[sor-1][oszlop-1]);


    // 3. Feladat
    println!("3. feladat:\nVilágos képpontok száma: {}",
           pixels.iter().flatten()
           .filter(|p| {
              p.0 + p.1 + p.2 > 600
           }).count()
    );


    // 4. Feladat
    let min = pixels.iter().flatten()
        .map(|p| {
            p.0 + p.1 + p.2
        }).min().unwrap();

    println!("4. feladat:\nA legsötétebb pont RGB összege: {}", min);

    println!("A legsötétebb pixelek színe:");
    for i in pixels.iter().flatten().filter(|p| p.0 + p.1 + p.2 == min)
    { println!("{:?}", i); }

    // 6. Feladat
    println!("6. feladat:");
    let (e, _) = pixels.iter().enumerate()
        .filter(|(_, l)| hatar(l, 10))
        .next().unwrap();
    let (u, _) = pixels.iter().enumerate()
        .filter(|(_, l)| hatar(l, 10))
        .last().unwrap();
    println!("A felhő legfelső sora: {}", e + 1);
    println!("A felhő legalsó sora: {}", u + 1);

}

// 5. Feladat
fn hatar(sor: &Vec<RGB>, elteres: i32) -> bool {
     if 0 < sor.iter()
        .map(|p| p.2 )
        .zip(sor[1..sor.len()].iter().map(|p| p.2))
        .filter(|(a, b)| (a-b).abs() > elteres)
        .count()
    { return true; }

    false
}
